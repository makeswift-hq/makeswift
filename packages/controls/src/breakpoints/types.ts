import { type Device as DeviceId } from '../common/types'

export type BreakpointId = DeviceId

export type Breakpoint = {
  id: BreakpointId
  label?: string
  viewportWidth?: number
  minWidth?: number
  maxWidth?: number
}

export type Breakpoints = Breakpoint[]
