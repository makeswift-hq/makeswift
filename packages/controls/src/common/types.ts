import { z } from 'zod'

import * as Schema from './schema'

export { type BoxModel } from 'css-box-model'

export type Data = z.infer<typeof Schema.data>
export type ColorData = z.infer<typeof Schema.colorData>
export type ResolvedColorData = z.infer<typeof Schema.resolvedColorData>
export type Device = z.infer<typeof Schema.deviceId>
export type DeviceOverride<T> = { deviceId: Device; value: T }
export type ResponsiveValue<T> = DeviceOverride<T>[]
export type ResponsiveValueType<T> =
  T extends ResponsiveValue<infer U> ? U : never

export type ElementData = z.infer<typeof Schema.elementData>
export type ElementReference = z.infer<typeof Schema.elementReference>
export type Element = z.infer<typeof Schema.element>

export function isElementReference(
  element: Element,
): element is ElementReference {
  return !('props' in element)
}

export const ControlDataTypeKey = '@@makeswift/type'
