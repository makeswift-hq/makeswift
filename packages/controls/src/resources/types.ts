import { z } from 'zod'

import * as Schema from './schema'

export type Swatch = z.infer<typeof Schema.swatch>
export type ColorData = z.infer<typeof Schema.colorData>
export type ResolvedColorData = z.infer<typeof Schema.resolvedColorData>
export type File = z.infer<typeof Schema.file>
export type PagePathnameSlice = z.infer<typeof Schema.pagePathnameSlice>
export type Typography = z.infer<typeof Schema.typography>
