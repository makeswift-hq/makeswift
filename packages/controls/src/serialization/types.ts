import { z } from 'zod'

import * as Schema from './schema'

export type SerializedRecord = z.infer<typeof Schema.serializedRecord>
export type DeserializedRecord = z.infer<typeof Schema.deserializedRecord>
