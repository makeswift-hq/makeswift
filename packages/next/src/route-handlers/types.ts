export type MakeswiftRouteHandlerConfig = {
  apiKey: string
  appOrigin?: string
  apiOrigin?: string
  unstable_siteVersions?: boolean
}
