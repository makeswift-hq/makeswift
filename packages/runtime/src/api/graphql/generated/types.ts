export type Maybe<T> = T | null
export type InputMaybe<T> = Maybe<T>
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] }

type Json = null | boolean | number | string | Json[] | { [key: string]: Json }

type Scalars = {
  ID: string
  JSON: Json
  Locale: string
}

export type CreateTableRecordInput = {
  data: CreateTableRecordInputData
}

export type CreateTableRecordInputData = {
  columns: Array<CreateTableRecordInputDataColumn>
  tableId?: InputMaybe<Scalars['ID']>
}

export type CreateTableRecordInputDataColumn = {
  columnId: Scalars['ID']
  data: Scalars['JSON']
}

export const SnippetLocation = {
  Body: 'BODY',
  Head: 'HEAD',
} as const

export type SnippetLocation = typeof SnippetLocation[keyof typeof SnippetLocation]

export type SwatchFragment = {
  __typename: 'Swatch'
  id: string
  hue: number
  saturation: number
  lightness: number
}

export type FileFragment = {
  __typename: 'File'
  id: string
  name: string
  extension: string | null
  publicUrl: string
  dimensions: { width: number; height: number } | null
}

export type TypographyFragment = {
  __typename: 'Typography'
  id: string
  name: string
  style: Array<{
    deviceId: string
    value: {
      fontFamily: string | null
      lineHeight: number | null
      letterSpacing: number | null
      fontWeight: number | null
      textAlign: string | null
      uppercase: boolean | null
      underline: boolean | null
      strikethrough: boolean | null
      italic: boolean | null
      fontSize: { value: number | null; unit: string | null } | null
      color: { swatchId: string | null; alpha: number | null } | null
    }
  }>
}

export type PagePathnameSliceFragment = {
  __typename: 'PagePathnameSlice'
  id: string
  pathname: string
}

export type GlobalElementFragment = { __typename: 'GlobalElement'; id: string; data: Json }

export type LocalizedGlobalElementFragment = {
  __typename: 'LocalizedGlobalElement'
  id: string
  data: Json
}

export type TableFragment = {
  __typename: 'Table'
  id: string
  name: string
  columns: Array<
    | { __typename: 'CheckboxTableColumn'; id: string; name: string }
    | { __typename: 'EmailTableColumn'; id: string; name: string }
    | { __typename: 'LongTextTableColumn'; id: string; name: string }
    | {
        __typename: 'MultipleSelectTableColumn'
        id: string
        name: string
        options: Array<{ id: string; name: string }>
      }
    | { __typename: 'NumberTableColumn'; id: string; name: string }
    | { __typename: 'PhoneNumberTableColumn'; id: string; name: string }
    | { __typename: 'SingleLineTextTableColumn'; id: string; name: string }
    | {
        __typename: 'SingleSelectTableColumn'
        id: string
        name: string
        options: Array<{ id: string; name: string }>
      }
    | { __typename: 'URLTableColumn'; id: string; name: string }
  >
}

export type SnippetFragment = {
  __typename: 'Snippet'
  id: string
  name: string
  code: string
  cleanup: string | null
  location: SnippetLocation
  shouldAddToNewPages: boolean
  liveEnabled: boolean
  builderEnabled: boolean
}

export type PageFragment = {
  __typename: 'Page'
  id: string
  snippets: Array<{
    __typename: 'Snippet'
    id: string
    name: string
    code: string
    cleanup: string | null
    location: SnippetLocation
    shouldAddToNewPages: boolean
    liveEnabled: boolean
    builderEnabled: boolean
  }>
}

export type SiteFragment = {
  __typename: 'Site'
  id: string
  googleFonts: {
    edges: Array<{
      activeVariants: Array<{ specifier: string }>
      node: { family: string; variants: Array<{ specifier: string }> }
    } | null>
  }
}

export type IntrospectedResourcesQueryVariables = Exact<{
  fileIds: Array<Scalars['ID']> | Scalars['ID']
  pageIds: Array<Scalars['ID']> | Scalars['ID']
  tableIds: Array<Scalars['ID']> | Scalars['ID']
}>

export type IntrospectedResourcesQueryResult = {
  files: Array<{
    __typename: 'File'
    id: string
    name: string
    extension: string | null
    publicUrl: string
    dimensions: { width: number; height: number } | null
  } | null>
  pagePathnamesById: Array<{ __typename: 'PagePathnameSlice'; id: string; pathname: string } | null>
  tables: Array<{
    __typename: 'Table'
    id: string
    name: string
    columns: Array<
      | { __typename: 'CheckboxTableColumn'; id: string; name: string }
      | { __typename: 'EmailTableColumn'; id: string; name: string }
      | { __typename: 'LongTextTableColumn'; id: string; name: string }
      | {
          __typename: 'MultipleSelectTableColumn'
          id: string
          name: string
          options: Array<{ id: string; name: string }>
        }
      | { __typename: 'NumberTableColumn'; id: string; name: string }
      | { __typename: 'PhoneNumberTableColumn'; id: string; name: string }
      | { __typename: 'SingleLineTextTableColumn'; id: string; name: string }
      | {
          __typename: 'SingleSelectTableColumn'
          id: string
          name: string
          options: Array<{ id: string; name: string }>
        }
      | { __typename: 'URLTableColumn'; id: string; name: string }
    >
  } | null>
}

export type SwatchQueryVariables = Exact<{
  swatchId: Scalars['ID']
}>

export type SwatchQueryResult = {
  swatch: {
    __typename: 'Swatch'
    id: string
    hue: number
    saturation: number
    lightness: number
  } | null
}

export type FileQueryVariables = Exact<{
  fileId: Scalars['ID']
}>

export type FileQueryResult = {
  file: {
    __typename: 'File'
    id: string
    name: string
    extension: string | null
    publicUrl: string
    dimensions: { width: number; height: number } | null
  } | null
}

export type TypographyQueryVariables = Exact<{
  typographyId: Scalars['ID']
}>

export type TypographyQueryResult = {
  typography: {
    __typename: 'Typography'
    id: string
    name: string
    style: Array<{
      deviceId: string
      value: {
        fontFamily: string | null
        lineHeight: number | null
        letterSpacing: number | null
        fontWeight: number | null
        textAlign: string | null
        uppercase: boolean | null
        underline: boolean | null
        strikethrough: boolean | null
        italic: boolean | null
        fontSize: { value: number | null; unit: string | null } | null
        color: { swatchId: string | null; alpha: number | null } | null
      }
    }>
  } | null
}

export type PagePathnamesByIdQueryVariables = Exact<{
  pageIds: Array<Scalars['ID']> | Scalars['ID']
}>

export type PagePathnamesByIdQueryResult = {
  pagePathnamesById: Array<{ __typename: 'PagePathnameSlice'; id: string; pathname: string } | null>
}

export type TableQueryVariables = Exact<{
  tableId: Scalars['ID']
}>

export type TableQueryResult = {
  table: {
    __typename: 'Table'
    id: string
    name: string
    columns: Array<
      | { __typename: 'CheckboxTableColumn'; id: string; name: string }
      | { __typename: 'EmailTableColumn'; id: string; name: string }
      | { __typename: 'LongTextTableColumn'; id: string; name: string }
      | {
          __typename: 'MultipleSelectTableColumn'
          id: string
          name: string
          options: Array<{ id: string; name: string }>
        }
      | { __typename: 'NumberTableColumn'; id: string; name: string }
      | { __typename: 'PhoneNumberTableColumn'; id: string; name: string }
      | { __typename: 'SingleLineTextTableColumn'; id: string; name: string }
      | {
          __typename: 'SingleSelectTableColumn'
          id: string
          name: string
          options: Array<{ id: string; name: string }>
        }
      | { __typename: 'URLTableColumn'; id: string; name: string }
    >
  } | null
}

export type TypographiesQueryVariables = Exact<{
  typographyIds: Array<Scalars['ID']> | Scalars['ID']
}>

export type TypographiesQueryResult = {
  typographies: Array<{
    __typename: 'Typography'
    id: string
    name: string
    style: Array<{
      deviceId: string
      value: {
        fontFamily: string | null
        lineHeight: number | null
        letterSpacing: number | null
        fontWeight: number | null
        textAlign: string | null
        uppercase: boolean | null
        underline: boolean | null
        strikethrough: boolean | null
        italic: boolean | null
        fontSize: { value: number | null; unit: string | null } | null
        color: { swatchId: string | null; alpha: number | null } | null
      }
    }>
  } | null>
}

export type GlobalElementQueryVariables = Exact<{
  globalElementId: Scalars['ID']
}>

export type GlobalElementQueryResult = {
  globalElement: { __typename: 'GlobalElement'; id: string; data: Json } | null
}

export type CreateTableRecordMutationVariables = Exact<{
  input: CreateTableRecordInput
}>

export type CreateTableRecordMutationResult = { createTableRecord: { tableRecord: { id: string } } }
