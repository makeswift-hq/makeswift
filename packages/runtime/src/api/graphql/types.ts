import {
  SwatchFragment as Swatch,
  FileFragment as File,
  TypographyFragment as Typography,
  PagePathnameSliceFragment as PagePathnameSlice,
  GlobalElementFragment as GlobalElement,
  LocalizedGlobalElementFragment as LocalizedGlobalElement,
  TableFragment as Table,
  SnippetFragment as Snippet,
  PageFragment as Page,
  SiteFragment as Site,
} from './generated/types'

export type {
  Swatch,
  File,
  Typography,
  PagePathnameSlice,
  GlobalElement,
  LocalizedGlobalElement,
  Table,
  Snippet,
  Page,
  Site,
}

export type APIResource =
  | Swatch
  | File
  | Typography
  | PagePathnameSlice
  | GlobalElement
  | LocalizedGlobalElement
  | Table
  | Snippet
  | Page
  | Site

export const APIResourceType: { [R in APIResource as R['__typename']]: R['__typename'] } = {
  Swatch: 'Swatch',
  File: 'File',
  Typography: 'Typography',
  PagePathnameSlice: 'PagePathnameSlice',
  GlobalElement: 'GlobalElement',
  LocalizedGlobalElement: 'LocalizedGlobalElement',
  Table: 'Table',
  Snippet: 'Snippet',
  Page: 'Page',
  Site: 'Site',
}

export type APIResourceType = typeof APIResourceType[keyof typeof APIResourceType]
