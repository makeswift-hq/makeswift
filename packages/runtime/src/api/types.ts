import {
  SwatchFragment as Swatch,
  FileFragment as File,
  TypographyFragment as Typography,
  PagePathnameSliceFragment as PagePathnameSlice,
  GlobalElementFragment as GlobalElement,
  TableFragment as Table,
  SnippetFragment as Snippet,
  PageFragment as Page,
} from './generated/graphql'

export type APIResource =
  | Swatch
  | File
  | Typography
  | PagePathnameSlice
  | GlobalElement
  | Table
  | Snippet
  | Page

export type APIResourceType = APIResource['__typename']
