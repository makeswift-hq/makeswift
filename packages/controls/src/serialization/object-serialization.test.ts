import {
  deserializeObject,
  Serializable,
  serializeObject,
  type Deserialized,
} from './object-serialization'
import { type SerializedRecord } from './types'

expect.addSnapshotSerializer({
  serialize: () => '[MessagePort]',
  test: (value) => value instanceof MessagePort,
})

class SerializableObject extends Serializable {
  constructor(private readonly obj: { type: string; [key: string]: any }) {
    super()
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serializeObject(this.obj) as [SerializedRecord, Transferable[]]
  }
}

describe('object serialization', () => {
  test('Deserialized', () => {
    const obj = {
      label: 'hello',
      getItemLabel: (item: { label: string }) => item.label,
    }

    const deserialized: Deserialized<typeof obj> = {
      label: 'hello',
      getItemLabel: async (item: { label: string }) => item.label,
    }

    expect(deserialized.getItemLabel).toMatchSnapshot()
  })

  describe.each([
    {},
    { label: undefined },
    { label: 'visible', defaultValue: true },
    { label: 'visible', options: [{ label: 'one', value: 1 }] },
  ])('%o', (object) => {
    test('should serialize to itself', () => {
      expect(serializeObject(object)).toEqual([object, []])
    })

    test('should deserialize to itself', () => {
      const [serialized, _] = serializeObject(object)
      expect(deserializeObject(serialized)).toEqual(object)
    })
  })

  test('should correctly serialize nested Serializables', () => {
    const object = {
      item: new SerializableObject({
        type: 'item',
        getItemLabel: (item: any) => item.label,
        subitems: [
          new SerializableObject({
            type: 'subitem 1',
            arrayOfFunctions: [(item: any) => item.label],
          }),
          new SerializableObject({
            type: 'subitem 2',
            getItemLabel: (item: any) => item.label,
          }),
        ],
      }),
    }

    const [serialized, transferables] = serializeObject(object)
    expect(transferables).toHaveLength(3)
    expect(transferables[0]).toBeInstanceOf(MessagePort)
    transferables.forEach((port: any) => port.close())

    expect(serialized).toMatchSnapshot()
  })

  describe.each([
    { getItemLabel: (item: any) => item.label },
    { arrayOfFunctions: [(item: any) => item.label] },
  ])('%o', (object) => {
    test('should handle function serialization', async () => {
      const [serialized, transferables] = serializeObject(object)
      expect(transferables).toHaveLength(1)
      expect(transferables[0]).toBeInstanceOf(MessagePort)
      transferables.forEach((port: any) => port.close())

      expect(serialized).toMatchSnapshot()
    })

    test('should deserialize serialized functions', () => {
      const [serialized, transferables] = serializeObject(object)
      transferables.forEach((port: any) => port.close())
      expect(deserializeObject(serialized)).toMatchSnapshot()
    })
  })
})
