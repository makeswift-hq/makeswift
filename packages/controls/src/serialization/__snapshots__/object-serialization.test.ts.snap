// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object serialization {
  arrayOfFunctions: [ [Function (anonymous)] { [length]: 1, [name]: '' }, [length]: 1 ]
} should deserialize serialized functions 1`] = `
{
  "arrayOfFunctions": [
    [Function],
  ],
}
`;

exports[`object serialization {
  arrayOfFunctions: [ [Function (anonymous)] { [length]: 1, [name]: '' }, [length]: 1 ]
} should handle function serialization 1`] = `
{
  "arrayOfFunctions": [
    [MessagePort],
  ],
}
`;

exports[`object serialization {
  getItemLabel: [Function: getItemLabel] { [length]: 1, [name]: 'getItemLabel' }
} should deserialize serialized functions 1`] = `
{
  "getItemLabel": [Function],
}
`;

exports[`object serialization {
  getItemLabel: [Function: getItemLabel] { [length]: 1, [name]: 'getItemLabel' }
} should handle function serialization 1`] = `
{
  "getItemLabel": [MessagePort],
}
`;

exports[`object serialization Deserialized 1`] = `[Function]`;

exports[`object serialization should correctly serialize nested Serializables 1`] = `
{
  "item": {
    "getItemLabel": [MessagePort],
    "subitems": [
      {
        "arrayOfFunctions": [
          [MessagePort],
        ],
        "type": "subitem 1",
      },
      {
        "getItemLabel": [MessagePort],
        "type": "subitem 2",
      },
    ],
    "type": "item",
  },
}
`;
